Data/ByteString.hs

```diff
@@ -978,28 +993,29 @@ spanEnd  p ps = splitAt (findFromEndUntil (not.p) ps) ps
 --
 splitWith :: (Word8 -> Bool) -> ByteString -> [ByteString]
 splitWith _pred (BS _  0) = []
-splitWith pred_ (BS fp len) = splitWith0 pred# 0 len fp
-  where pred# c# = pred_ (W8# c#)
+splitWith pred_ (BS fp len) = splitWith0 0 len fp
+  where pred# :: (Word# -> Bool)
+        pred# c# = pred_ (W8# c#)
 
-        splitWith0 !pred' !off' !len' !fp' =
+        splitWith0 !off' !len' !fp' =
           accursedUnutterablePerformIO $
             withForeignPtr fp $ \p ->
-              splitLoop pred' p 0 off' len' fp'
+              splitLoop p 0 off' len' fp'
 
-        splitLoop :: (Word# -> Bool)
-                  -> Ptr Word8
+        splitLoop :: Ptr Word8
                   -> Int -> Int -> Int
                   -> ForeignPtr Word8
                   -> IO [ByteString]
-
-        splitLoop pred' p idx' off' len' fp'
-            | idx' >= len'  = return [BS (plusForeignPtr fp' off') idx']
-            | otherwise = do
-                w <- peekElemOff p (off'+idx')
-                if pred' (case w of W8# w# -> w#)
-                   then return (BS (plusForeignPtr fp' off') idx' :
-                              splitWith0 pred' (off'+idx'+1) (len'-idx'-1) fp')
-                   else splitLoop pred' p (idx'+1) off' len' fp'
+        splitLoop p idx2 off' len' fp' = go idx2
+          where
+            go idx'
+                | idx' >= len'  = return [BS (plusForeignPtr fp' off') idx']
+                | otherwise = do
+                    w <- peekElemOff p (off'+idx')
+                    if pred# (case w of W8# w# -> w#)
```

Comment:
I'm not sure what purpose the use of `pred#` is intended to serve, but if it is actually useful, why not:
```suggestion
                    !(W8# w#) <- peekElemOff p (off'+idx')
                    if pred# (W8# w#)
```
Which should be equivalent, but what's the point here?  I'm rather unclear on what any of this is doing, are we somehow helping the optimizer evaluate equality predicates more efficiently here?  Wouldn't that happen without this?

Reply:
At a glance I do not see the point either. But since `pred#` remained unchanged in course of this PR, I'd rather postpone the investigation until another occasion.

Reply:
Fair enough, but it still looks rather questionable to me.  I'd like to see it eliminated, and see if that makes things any worse (it might make things better in fact).

Reply:
I checked that Core for `pred# (case w of W8# w# -> w#)` and `pred w` is similar, so there is not reason to resort to `MagicHash` here. Thanks for the hint!

Reply:
Thanks for confirming my intuition, that helps. I just couldn't imagine how the original code could be a win, and happy to see it wasn't "for lack of imagination"...